engine/src/test/java/io/camunda/zeebe/engine/processing/bpmn/multiinstance/MultiInstanceActivityTest.java

```diff
@@ -265,6 +266,40 @@ public void shouldCompleteBodyWhenAllJobsAreCompleted() {
             tuple(BpmnElementType.MULTI_INSTANCE_BODY, ProcessInstanceIntent.ELEMENT_COMPLETED));
   }
 
+  @Test
+  public void shouldCompleteBodyWithCompleteCondition() {
+    // given
+    ENGINE
+        .deployment()
+        .withXmlResource(
+            process(miBuilder.andThen(m -> m.completionCondition(COMPLETION_CONDITION_EXPRESSION))))
+        .deploy();
+
+    // when
+    final long processInstanceKey =
+        ENGINE
+            .processInstance()
+            .ofBpmnProcessId(PROCESS_ID)
+            .withVariable(INPUT_COLLECTION_EXPRESSION, INPUT_COLLECTION)
+            .create();
+
+    final int completedJobs = 1;
+    completeJobs(processInstanceKey, completedJobs);
+
+    // then
+    assertThat(
+            RecordingExporter.processInstanceRecords()
+                .withProcessInstanceKey(processInstanceKey)
+                .limitToProcessInstanceCompleted()
+                .withElementId(ELEMENT_ID))
+        .extracting(r -> tuple(r.getValue().getBpmnElementType(), r.getIntent()))
+        .containsSubsequence(
+            tuple(BpmnElementType.SERVICE_TASK, ProcessInstanceIntent.ELEMENT_COMPLETED),
+            tuple(BpmnElementType.MULTI_INSTANCE_BODY, ProcessInstanceIntent.COMPLETE_ELEMENT),
+            tuple(BpmnElementType.MULTI_INSTANCE_BODY, ProcessInstanceIntent.ELEMENT_COMPLETING),
+            tuple(BpmnElementType.MULTI_INSTANCE_BODY, ProcessInstanceIntent.ELEMENT_COMPLETED));
+  }
```

Comment:
🔧   I think this should also check that the other activated service tasks were terminated in the parallel multi-instance **and** that no other service tasks activated for the sequential multi-instance.

Something like:
```java
if (loopCharacteristics.equals("parallel")) {
      // after 1 has completed, the others must be terminated
      final var expectedNumberOfTerminatedServiceTasks = 3 - 1;
      assertThat(
              RecordingExporter.records()
                  .limitToProcessInstance(processInstanceKey)
                  .processInstanceRecords()
                  .withIntent(ProcessInstanceIntent.ELEMENT_TERMINATED)
                  .withElementType(BpmnElementType.SERVICE_TASK)
                  .count())
          .describedAs("all non-completed service tasks have terminated")
          .isEqualTo(expectedNumberOfTerminatedServiceTasks);
    } else {
      assertThat(
          RecordingExporter.records()
              .limitToProcessInstance(processInstanceKey)
              .processInstanceRecords()
              .withIntent(ProcessInstanceIntent.ELEMENT_ACTIVATED)
              .withElementType(BpmnElementType.SERVICE_TASK)
              .count())
          .describedAs("only 1 out of 3 sequential service tasks has activated")
          .isEqualTo(1);
    }
```

Reply:
>   it should start terminating all the active child instances (by writing the TERMINATE_ELEMENT commands for them), which will `asynchronously` start terminating those tasks (and as a consequence cancel their respective jobs)

Hi @korthout , thank you for responding so quickly!  As you said, at the beginning I did use the `afterExecutionPathCompleted` method to terminate all active instances.

But in a special scenario, I encountered a problem:

If there are 3 instances and only two tasks are need completed according to the completion condtion expression, the last job should be  cancel immediately. When client batch activates jobs, because our processor terminates it asynchronously, and the client handle job in multi-threads, so occasionally all job will be completed. 

It is precisely because of this that I put terminate the logic of the job  into the `JobCompleteProcessor` for processing. When the `completionCondition` evaluates to true, all active element instances are immediately terminated, as are the modifications of other processors.

Reply:
>If there are 3 instances and only two tasks are need completed according to the completion condtion expression, the last job should be cancel immediately. When client batch activates jobs, because our processor terminates it asynchronously, and the client handle job in multi-threads, so occasionally all job will be completed.

You make an excellent point. However, I wouldn't worry about that right now.

Zeebe already uses this same asynchronous logic for process instance cancelation and other interruptions (like interrupting catch events and such). This is because the part we want to terminate can contain complex structures, which is also the case for a multi-instance element. For example, let's consider a multi-instance call activity that calls another process. We should not just terminate some jobs directly, but terminate the called child process instance and wait until it has completely terminated. The termination of the child process instance will terminate its children recursively. If we were to do try to implement all of that here, we would also have to deal with unsubscribing of message, timer and error event subscriptions, and possibly other concepts. It makes things much more complicated. 

Let's make things simpler by just terminating the child instances asynchronously. If we do have a need to restrict a job from being activatable immediately after the interrupt then we can implement something specific for that later.

Reply:
Hi @korthout , I've fixed review comments, please check this out. :)